src/sys/socket/addr.rs

```diff
@@ -1761,6 +1761,89 @@ pub mod netlink {
     }
 }
 
+#[cfg(target_os = "linux")]
+#[cfg_attr(docsrs, doc(cfg(all())))]
+pub mod can {
+    use super::*;
+    use crate::sys::socket::addr::AddressFamily;
+    use libc::{sa_family_t, sockaddr_can, c_int, c_uint};
+    use std::{fmt, mem};
+
+    /// Address for the Linux SocketCAN implementation for the Controller Area Network
+    /// automotive bus protocol.
+    ///
+    /// # References
+    ///
+    /// [SocketCAN](https://docs.kernel.org/networking/can.html)
+    #[derive(Copy, Clone)]
+    #[allow(missing_debug_implementations)]
+    #[repr(transparent)]
+    pub struct CanAddr(pub(in super::super) sockaddr_can);
+
+    impl CanAddr {
+        /// Construct a new CAN socket address from its interface index.
+        pub fn new(ifindex: c_uint) -> Self {
+            let mut addr: sockaddr_can = unsafe { mem::zeroed() };
+            addr.can_family = AddressFamily::Can as sa_family_t;
+            addr.can_ifindex = ifindex as c_int;
+
+            Self(addr)
+        }
+
+        /// Return the socket's CAN interface index.
+        pub const fn ifindex(&self) -> c_uint {
```

Comment:
TLDR: A valid index can only be >=0 (zero for "bind all" and >=1 for a specific index). The `new()` function takes a uint and the getter returns that uint as the same type. The weird this is that  the `sockaddr_can` struct takes it as signed int, though a valid one can never be less than zero. So keeping the public API as uint seems the most consistent.

More details :

This is actually a weird one. 

The C function [if_nametoindex](https://man7.org/linux/man-pages/man3/if_nametoindex.3.html) is used to look up the index number from a name (like "can0", "can1", "vcan0", etc). Very strangely, it returns an `unsigned` with zero indicating an error (instead of the normal C API convention of returning -1 on error) and >=1 giving a valid index number.

The nix call to get the interface number, [if_nametoindex](https://docs.rs/nix/latest/nix/net/if_/fn.if_nametoindex.html) returns a `Result<c_uint>`, which makes perfect sense.

But then, for some reason, the `sockaddr_can` struct takes a signed `int` vlaue (`c_int`). But it really can only be an int >= 1.